src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.cpp

```diff
@@ -0,0 +1,974 @@
+
+bool ShenandoahDirectCardMarkRememberedSet::is_write_card_dirty(HeapWord* p) const {
+  size_t index = card_index_for_addr(p);
+  CardValue* bp = &(_card_table->write_byte_map())[index];
+  return (bp[0] == CardTable::dirty_card_val());
+}
+
+void ShenandoahDirectCardMarkRememberedSet::mark_card_as_dirty(HeapWord* p) {
+  size_t index = card_index_for_addr(p);
+  CardValue* bp = &(_card_table->write_byte_map())[index];
+  bp[0] = CardTable::dirty_card_val();
+}
+
+void ShenandoahDirectCardMarkRememberedSet::mark_range_as_dirty(HeapWord* p, size_t num_heap_words) {
+  CardValue* bp = &(_card_table->write_byte_map_base())[uintptr_t(p) >> _card_shift];
+  CardValue* end_bp = &(_card_table->write_byte_map_base())[uintptr_t(p + num_heap_words) >> _card_shift];
+  // If (p + num_heap_words) is not aligned on card boundary, we also need to dirty last card.
+  if (((unsigned long long) (p + num_heap_words)) & (CardTable::card_size() - 1)) {
+    end_bp++;
+  }
+  while (bp < end_bp) {
+    *bp++ = CardTable::dirty_card_val();
+  }
+}
+
+void ShenandoahDirectCardMarkRememberedSet::mark_card_as_clean(HeapWord* p) {
+  size_t index = card_index_for_addr(p);
+  CardValue* bp = &(_card_table->write_byte_map())[index];
+  bp[0] = CardTable::clean_card_val();
+}
+
+void ShenandoahDirectCardMarkRememberedSet::mark_range_as_clean(HeapWord* p, size_t num_heap_words) {
+  CardValue* bp = &(_card_table->write_byte_map_base())[uintptr_t(p) >> _card_shift];
+  CardValue* end_bp = &(_card_table->write_byte_map_base())[uintptr_t(p + num_heap_words) >> _card_shift];
+  // If (p + num_heap_words) is not aligned on card boundary, we also need to clean last card.
+  if (((unsigned long long) (p + num_heap_words)) & (CardTable::card_size() - 1)) {
+    end_bp++;
+  }
+  while (bp < end_bp) {
+    *bp++ = CardTable::clean_card_val();
+  }
+}
+
+// No lock required because arguments align with card boundaries.
+void ShenandoahCardCluster::reset_object_range(HeapWord* from, HeapWord* to) {
+  assert(((((unsigned long long) from) & (CardTable::card_size() - 1)) == 0) &&
+         ((((unsigned long long) to) & (CardTable::card_size() - 1)) == 0),
+         "reset_object_range bounds must align with card boundaries");
+  size_t card_at_start = _rs->card_index_for_addr(from);
+  size_t num_cards = (to - from) / CardTable::card_size_in_words();
+
+  for (size_t i = 0; i < num_cards; i++) {
+    _object_starts[card_at_start + i].short_word = 0;
+  }
+}
+
+// Assume only one thread at a time registers objects pertaining to
+// each card-table entry's range of memory.
+void ShenandoahCardCluster::register_object(HeapWord* address) {
+  shenandoah_assert_heaplocked();
+
+  register_object_without_lock(address);
+}
+
+void ShenandoahCardCluster::register_object_without_lock(HeapWord* address) {
+  size_t card_at_start = _rs->card_index_for_addr(address);
+  HeapWord* card_start_address = _rs->addr_for_card_index(card_at_start);
+  uint8_t offset_in_card = address - card_start_address;
```

Comment:
Sonar: implicit conversion loses integer precision: 'long' to 'uint8_t' (aka 'unsigned char').

This looks risky. Should probably be `checked_cast<uint8_t>(pointer_delta(address, card_start_address, 1))`?

Reply:
https://bugs.openjdk.org/browse/JDK-8344321